src/classlibnative/bcltype/number.cpp

```diff
@@ -137,86 +137,185 @@ unsigned int Int64DivMod1E9(unsigned __int64* value)
 
 void ecvt( double value, int count, int* dec, int* sign, wchar_t* digits )
 {
+    // ========================================================================================================================================
+    // This implementation is based on the paper: https://www.cs.indiana.edu/~dyb/pubs/FP-Printing-PLDI96.pdf
+    // Besides the paper, some of the code and ideas are modified from http://www.ryanjuckett.com/programming/printing-floating-point-numbers/
+    // You must read these two marterials to fully understand the code.
+    //
+    // Note: we only support fixed format input.
+    // ======================================================================================================================================== 
+    //
+    // Overview:
+    //
+    // The input double number can be represented as:
+    // value = f * 2^e = r / s.
+    //
+    // f: the output mantissa. Note: f is not the 52 bits mantissa of the input double number. 
+    // e: biased exponent.
+    // r: numerator.
+    // s: denominator.
+    // k: value = d0.d1d2 . . . dn * 10^k
+
+    _ASSERTE(dec != nullptr && sign != nullptr && digits != nullptr);
+
+    // The caller of _ecvt should already checked the Infinity and NAN values.
+    _ASSERTE(((FPDOUBLE*)&value)->exp != 0x7ff);
+
+    // Shortcut for zero.
+    if (value == 0.0)
+    {
+        *dec = 0;
+        *sign = 0;
+
+        for (int i = 0; i < count; ++i)
+        {
+            digits[i] = L'0';
+        }
+
+        digits[count] = 0;
+
+        return;
+    } 
+
     // Step 1:
     // Extract meta data from the input double value.
     //
     // Refer to IEEE double precision floating point format.
-    UINT64 realMantissa = 0;
-    int realExponent = 0;
+    UINT64 f = 0;
+    int e = 0;
     UINT32 mantissaHighBitIdx = 0;
     if (((FPDOUBLE*)&value)->exp > 0)
     {
-        realMantissa = ((UINT64)(((FPDOUBLE*)&value)->mantHi) << 32) | ((FPDOUBLE*)&value)->mantLo + ((UINT64)1 << 52);
-        realExponent = ((FPDOUBLE*)&value)->exp - 1075;
+        // For normalized value, according to https://en.wikipedia.org/wiki/Double-precision_floating-point_format
+        // value = 1.fraction * 2^(exp - 1023) 
+        //       = (1 + mantissa / 2^52) * 2^(exp - 1023) 
+        //       = (2^52 + mantissa) * 2^(exp - 1023 - 52)
+        //
+        // So f = (2^52 + mantissa), e = exp - 1075; 
+        f = ((UINT64)(((FPDOUBLE*)&value)->mantHi) << 32) | ((FPDOUBLE*)&value)->mantLo + ((UINT64)1 << 52);
+        e = ((FPDOUBLE*)&value)->exp - 1075;
         mantissaHighBitIdx = 52;
     }
     else
     {
-        realMantissa = ((UINT64)(((FPDOUBLE*)&value)->mantHi) << 32) | ((FPDOUBLE*)&value)->mantLo;
-        realExponent = -1074;
-        mantissaHighBitIdx = BigNum::LogBase2(realMantissa);
+        // For denormalized value, according to https://en.wikipedia.org/wiki/Double-precision_floating-point_format
```

Comment:
I don't think this is just for `denormalized` values which is only `exp = 0`. Based on the above comparison for `exp > 0`, this will also cover normalized values with a negative exp.